### PR TITLE
[WIP] Adding SUPG to ASPECT

### PR DESCRIPTION
**Pull request by rrgrove6**
Rough draft of adding SUPG to ASPECT.

Attached is a picture (I think) that shows the comparison between my code (left) and ASPECT (right) and SUPG is falling a little short of expectations (I assume it has something to do with the residual, but the problem could be anywhere).  It is still a long way from being pretty, but I figured the sooner I get eyes on it to help, the better.

comparison

### CODE REVIEW

**Comment by rrgrove6** [comment]
So, changed the residual to match the strong form, but now the result is quite worse.. not sure what I missed though.

**Comment by rrgrove6** [comment]
p.s. ignore whatever weird GIT stuff that I did in between the lastest SUPG commit and this one.

**Comment by rrgrove6** [comment]
After applying Timo's suggested changes to the residual, we have the current situation:

<--- EV
SUPG --->

comparison

NOTE: There must still be something incorrect in the residual. I am exploring the relationship between field_term_for_rhs and factor, as well as reaction_term.  Any addtionals areas of concern or ideas are welcome (:

**Comment by rrgrove6** [comment]
Current state:
ev <--
supg -->

**Comment by tjhei** [comment]
Please:
1. rebase to master
2. squash commits
3. change test to have supg=true

**Comment by rrgrove6** [comment]
@tjhei both steinberger tests fail on my laptop but supg passes.

File: source/simulator/assembly.cc [hunk 1]
```diff
@@ -280,22 +282,23 @@ namespace aspect
...
-          old_old_pressure_gradients (quadrature.size(), Utilities::signaling_nan<Tensor<1,dim> >()),
-          old_strain_rates (quadrature.size(), Utilities::signaling_nan<SymmetricTensor<2,dim> >()),
-          old_old_strain_rates (quadrature.size(), Utilities::signaling_nan<SymmetricTensor<2,dim> >()),
-          old_temperature_values (quadrature.size(), Utilities::signaling_nan<double>()),
-          old_old_temperature_values(quadrature.size(), Utilities::signaling_nan<double>()),
-          old_field_grads(quadrature.size(), Utilities::signaling_nan<Tensor<1,dim> >()),
-          old_old_field_grads(quadrature.size(), Utilities::signaling_nan<Tensor<1,dim> >()),
-          old_field_laplacians(quadrature.size(), Utilities::signaling_nan<double>()),
-          old_old_field_laplacians(quadrature.size(), Utilities::signaling_nan<double>()),
+          phi_field (advection_element.dofs_per_cell),
```
Comment (tjhei) [inline source/simulator/assembly.cc]: why this change?

Comment (rrgrove6) [inline source/simulator/assembly.cc]: I believe it came from some merge with master somewhere along the line.  Should it be changed back?

Comment (bangerth) [inline source/simulator/assembly.cc]: Definitely. Of course, this suggests the question whether there are other parts of the patch that also overwrite recently introduced functionality? Did you read through your patch again?
